File: app/src/main/java/io/github/tobyhs/weatherweight/api/accuweather/forecasts/DailyForecast.java
package io.github.tobyhs.weatherweight.api.accuweather.forecasts;

import com.google.auto.value.AutoValue;
import com.google.gson.annotations.SerializedName;
import me.tatarka.gsonvalue.annotations.GsonBuilder;
import org.threeten.bp.ZonedDateTime;

/**
 * A forecast for a day from AccuWeather's Forecasts API
 */
@AutoValue
public abstract class DailyForecast {
    /**
     * @return the date that this forecast is for
     */
    @SerializedName("Date")
    public abstract ZonedDateTime getDate();

    /**
     * @return the forecasted minimum and maximum temperatures
     */
    @SerializedName("Temperature")
    public abstract TemperatureRange getTemperature();

    /**
     * @return forecast details about the day
     */
    @SerializedName("Day")
    public abstract DayPeriod getDay();

    /**
     * @return forecast details about the night
     */
    @SerializedName("Night")
    public abstract DayPeriod getNight();

    @GsonBuilder
    public static Builder builder() {
        return new AutoValue_DailyForecast.Builder();
    }

    @AutoValue.Builder
    public static abstract class Builder {
        public abstract Builder setDate(ZonedDateTime date);
        public abstract Builder setTemperature(TemperatureRange temperature);
        public abstract Builder setDay(DayPeriod day);
        public abstract Builder setNight(DayPeriod night);
        public abstract DailyForecast build();
    }
}

File: app/src/main/java/io/github/tobyhs/weatherweight/api/accuweather/forecasts/DailyForecast.kt
package io.github.tobyhs.weatherweight.api.accuweather.forecasts

import java.time.ZonedDateTime

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

/**
 * A forecast for a day from AccuWeather's Forecasts API
 *
 * @property date the date that this forecast is for
 * @property temperature the forecasted minimum and maximum temperatures
 * @property day forecast description about the day
 * @property night forecast description about the night
 */
@JsonClass(generateAdapter = true)
data class DailyForecast(
    @Json(name = "Date") val date: ZonedDateTime,
    @Json(name = "Temperature") val temperature: TemperatureRange,
    @Json(name = "Day") val day: DayPeriod,
    @Json(name = "Night") val night: DayPeriod,
)
